### DISCUSSION TITLE
multiple variables of computeResidual()

### DISCUSSION THREAD

**Discussion by dingqiushi** [idaholab/moose#21427] [Q&A General] [answered]
I want to write the `computeResidual() `function in Kernel.C in a way that is suitable for multiple variables, where p is the pth variable, and I want to know if` _re_blocks` is designed for multiple variables, if the number of rows or columns of the residual matrix is affected by the number of variables, and to accommodate multiple variables, " MooseIndex(_vector_tags) i = 0; i < _vector_tags.size(); i++, ++vector_tag" how to rewrite the loop statement?

```
for (unsigned int p = 0; p < _n_equation; ++p)
	{
		  _re_blocks.resize(_vector_tags.size());
		  mooseAssert(_vector_tags.size() >= 1, "we need at least one active tag");
		  auto vector_tag = _vector_tags.begin();
		  for (MooseIndex(_vector_tags) i = 0; i < _vector_tags.size(); i++, ++vector_tag)
		    _re_blocks[i] = &assembly.residualBlock(p, *vector_tag);
		  _local_re.resize(_re_blocks[0]->size());

		for (_qp = 0; _qp < _qrule->n_points(); _qp++)
		{
			precalculateResidual();
			for (_i = 0; _i < _test.size(); _i++)
				_local_re(_i)+=_JxW[_qp] * _coord[_qp] * computeQpResidual(p);
		}

	  for (auto & re : _re_blocks)
	    *re += _local_re;
	}
```

**Comment by YaqiWang** [comment, marked as answer]
You may want to check out `ArrayKernel.C`, which works with an array variable having multiple components. This link documents some useful info about array variables: https://mooseframework.inl.gov/source/variables/ArrayMooseVariable.html.
